### PR TITLE
Alerting: DingDing notification channel now includes alert values

### PR DESCRIPTION
**Pull request by athurg**
Add the match value('s) into Dingding's notification message.

First I just add the values into message. But it not work better. Because we always has many match values. The original code use only *link* msgtype, which will truncate the mass text.

So I add the msgtype scheleton support, with a new msgtype *actionCard*, which will never truncate anything.

```release-note
Dingding alert message now includes values and an option for a new actionCard type. 
```

### CODE REVIEW

**Comment by torkelo** [comment]
Thanks for contributing! Sorry for not reviewing this sooner. Could you show screenshots of how the notification now looks?

File: pkg/services/alerting/notifiers/dingding.go [hunk 1]
```diff
@@ -1,26 +1,35 @@
...
+const DingdingOptionsTemplate = `
       <h3 class="page-heading">DingDing settings</h3>
       <div class="gf-form">
         <span class="gf-form-label width-10">Url</span>
         <input type="text" required class="gf-form-input max-width-26" ng-model="ctrl.model.settings.url" placeholder="https://oapi.dingtalk.com/robot/send?access_token=xxxxxxxxx"></input>
       </div>
-    `,
+      <div class="gf-form">
+        <span class="gf-form-label width-10">MessageType</span>
+        <select class="gf-form-input max-width-14" ng-model="ctrl.model.settings.msgType" ng-options="s for s in ['link','actionCard']" ng-init="ctrl.model.settings.msgType=ctrl.model.settings.msgType || '` + DefaultDingdingMsgType + `'"></select>
```
Comment (torkelo) [inline pkg/services/alerting/notifiers/dingding.go]: Is there a reason to introduce a new option?

Would not all want this New behavior?

Comment (athurg) [reply pkg/services/alerting/notifiers/dingding.go]: It's used for choose the type of Dingding Message.

The default value you seen is `link`, which is compatible with old(current) version. So when people upgrade the the latest version, it won't change the old user's experience.

Another option is `actionCard`, which provide a better experience like the screenshot I post above.

Comment (torkelo) [reply pkg/services/alerting/notifiers/dingding.go]: but won't everyone want the new actionCard? Why keep the old link format?

Comment (athurg) [reply pkg/services/alerting/notifiers/dingding.go]: > but won't everyone want the new actionCard? Why keep the old link format?

Maybe someone does. Anyway, if you think it's not necessary, I can just remove that option.

**Comment by athurg** [comment]
I just add another option `OpenInBrowser`, which to let user decide where the message url will be open, outside of Dingding ( in OS's default browser ), or just keep inside of the Dingding.

File: pkg/services/alerting/notifiers/dingding.go [hunk 2]
```diff
@@ -61,15 +95,34 @@ func (this *DingDingNotifier) Notify(evalContext *alerting.EvalContext) error {
...
-	}`))
+	for i, match := range evalContext.EvalMatches {
+		message += fmt.Sprintf("\\n%2d. %s: %s", i+1, match.Metric, match.Value)
+	}
+
+	var bodyStr string
+	if this.MsgType == "actionCard" {
+		bodyStr = `{
+			"msgtype": "actionCard",
+			"actionCard": {
```
Comment (torkelo) [inline pkg/services/alerting/notifiers/dingding.go]: looks like the actionCard does not have a picUrl equivalent so changing to actionCard by default (or removing) link opton would be bad I guess

File: pkg/services/alerting/notifiers/dingding.go [hunk 3]
```diff
@@ -1,26 +1,41 @@
...
       </div>
-    `,
+      <div class="gf-form">
+        <span class="gf-form-label width-10">MessageType</span>
+        <select class="gf-form-input max-width-14" ng-model="ctrl.model.settings.msgType" ng-options="s for s in ['link','actionCard']" ng-init="ctrl.model.settings.msgType=ctrl.model.settings.msgType || '` + DefaultDingdingMsgType + `'"></select>
+      </div>
+      <div class="gf-form">
+        <span class="gf-form-label width-10">OpenInBrowser</span>
+        <gf-form-switch class="gf-form" checked="ctrl.model.settings.openInBrowser"></gf-form-switch>
+        <info-popover mode="right-normal">Open the message url in browser instead of inside of Dingding</info-popover>
```
Comment (torkelo) [inline pkg/services/alerting/notifiers/dingding.go]: IS this not what you always want?

Comment (athurg) [reply pkg/services/alerting/notifiers/dingding.go]: Not always, Someone maybe want open inside of dingding

**Comment by athurg** [comment]
any more problems?

**Comment by athurg** [comment]
TEN more days ago

File: pkg/services/alerting/notifiers/dingding.go [hunk 4]
```diff
@@ -61,15 +95,39 @@ func (this *DingDingNotifier) Notify(evalContext *alerting.EvalContext) error {
...
+				"title": "` + title + `",
+				"singleTitle": "More",
+				"singleURL": "` + messageUrl + `"
+			}
+		}`
+	} else {
+		bodyStr = `{
+			"msgtype": "link",
+			"link": {
+				"text": "` + message + `",
```
Comment (torkelo) [inline pkg/services/alerting/notifiers/dingding.go]: maybe write a small utility function to quote a string.  Is this really nesssary though? seems like a very strange json api to have to quote strings inside strings.

**Review by torkelo — CHANGES_REQUESTED** [review]
We would prefer to not have so many options, it's hard to evaluate if they are needed by more than one person. This was about changing the message type, then a new option was added for message type, then a completely unrelated change  & new option was added to this PR.  Made reviewing it and deciding on merging harder. Prefer that features start as feature request issues so we have a chance to review if it's a worthy feature to add based on interest. 

Beside the new options that we are not sure about the unit test would need to be updated as well: https://github.com/grafana/grafana/blob/919d00437e21944d5feea3c6ac175a2d85736784/pkg/services/alerting/notifiers/dingding_test.go

If you removed the openInBrowser option, and only had the msgType option + unit test, then we can merge.

**Comment by athurg** [comment]
> We would prefer to not have so many options, it's hard to evaluate if they are needed by more than one person. This was about changing the message type, then a new option was added for message type, then a completely unrelated change & new option was added to this PR. Made reviewing it and deciding on merging harder. Prefer that features start as feature request issues so we have a chance to review if it's a worthy feature to add based on interest.
> 
> Beside the new options that we are not sure about the unit test would need to be updated as well: https://github.com/grafana/grafana/blob/919d00437e21944d5feea3c6ac175a2d85736784/pkg/services/alerting/notifiers/dingding_test.go
> 
> If you removed the openInBrowser option, and only had the msgType option + unit test, then we can merge.

- I have remote the openInBrowser option. 
- I have checked the unit test case, which have already cover all the new code.
- The string quote have been added. The best way is add a pure struct to describe the API request body, and then use json from golang's standard package to marshal it. But the origin code use a string concat and a self-defined simplejson lib to do that. So I choose to follow the behavior, because I don't know if there's any special reason. If you agree with the same best way, I can make another PR to change that, but it won't be here. This PR was opened so many days, and should be closed.

File: pkg/services/alerting/notifiers/dingding.go (hunk 4 above)
Comment (athurg) [reply pkg/services/alerting/notifiers/dingding.go]: FIXED

**Comment by athurg** [comment]
Sorry for force-push to reset the commiter email.

**Comment by torkelo** [comment]
Thanks for contributing! Merged and will be included in v6.1

**Pull request merged**